[export] _PythonPrinter for export's runtime asserts

pianpwk opened the pull request:
Prettifies export's runtime asserts, printing expressions with the _PythonPrinter and replacing symbols with prettified sources, to avoid expressions like `Eq(s0*s1, s2)`.

One step this involves for non-strict export is replacing the sources with what they'd look like for strict mode - non-strict contains sources like `L["args"][0][..]` for args, and `L["args"][1][..]` for kwargs, while sources from strict just contain the arg/kwarg name (e.g. `L[<name>][..]`). These sources are used for prettifying in later steps that are independent of strict/non-strict (e.g. run_decompositions()), so we just do a one-time replacement for these sources in non-strict. Kind of hacky, but matches up strict/non-strict and we only worry about it once.

A printer is added to the runtime asserts pass, defaulting to SymExprPrinter.

Export's runtime asserts now look like:
```
"Runtime assertion failed for expression (x.size(0)//2) <= y.size(0)"
```

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -287,6 +287,61 @@ def produce_guards_and_solve_constraints(
+    if shape_env.dim_constraints is None:
+        return  # no dynamic shapes
+
+    def _replace_sources(source):
```
cc @angelayi who made a good point about not doing this because runtime asserts are on a graph module whose placeholder names are divorced from source already, so it might be better to just use the placeholder names here.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5602,6 +5602,11 @@ def _print_Relational(self, expr):
         rhs = self.parenthesize(expr.rhs, sympy.printing.precedence.precedence(expr))
         return f"{lhs} {rel_op} {rhs}"
 
+    def _print_Mod(self, expr):
```
cool

avikchaudhuri reviewed (commented):
Overall looking good, other than the decision to use source names in runtime assertions. It might be better to build a symbol -> placeholder mapping instead.

pianpwk committed:
[export] _PythonPrinter for export's runtime asserts (#132821)

Summary:
Prettifies export's runtime asserts, printing expressions with the _PythonPrinter and replacing symbols with prettified sources, to avoid expressions like `Eq(s0*s1, s2)`.

One step this involves for non-strict export is replacing the sources with what they'd look like for strict mode - non-strict contains sources like `L["args"][0][..]` for args, and `L["args"][1][..]` for kwargs, while sources from strict just contain the arg/kwarg name (e.g. `L[<name>][..]`). These sources are used for prettifying in later steps that are independent of strict/non-strict (e.g. run_decompositions()), so we just do a one-time replacement for these sources in non-strict. Kind of hacky, but matches up strict/non-strict and we only worry about it once.

A printer is added to the runtime asserts pass, defaulting to SymExprPrinter.

Export's runtime asserts now look like:
```
"Runtime assertion failed for expression (x.size(0)//2) <= y.size(0)"
```


Differential Revision: D60870418

ezyang commented on test/export/test_export.py:
```diff
@@ -2189,7 +2189,7 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         with self.assertRaisesRegex(
             RuntimeError,
-            r"Runtime assertion failed for expression Eq\(Mod\(s0\*s1, s0 \- 1\), 0\)",
+            r"Runtime assertion failed for expression \(\(x\.size\(0\)\*x\.size\(1\)\) % \(x\.size\(0\) - 1\)\) == 0",
```
Can you do better on parenthesization? Also, I would suggest converting this into assertExpectedInline tests, so it's easier to update the output strings when they wobble

ezyang commented on torch/_export/utils.py:
```diff
@@ -633,3 +633,41 @@ def remove_proxy_from_state_dict(state_dict: Dict, in_place: bool) -> Dict:
+    # be safe and default to symbol names for remaining symbols (includes unbacked)
+    for symbol in shape_env.var_to_range.keys():
+        if symbol.name not in src_map:
+            src_map[symbol.name] = [symbol.name]
```
I think I would appreciate debug log prints specifying the mapping, in case I did want the original s0 numbering because I need to correlate it with other logs

ezyang commented:
Endorsing the symbolic shapes changes, I'll let y'all figure out the export side
